rules/prevent-abbreviations.js

```diff
@@ -0,0 +1,369 @@
+
+	// A `class` declaration produces two variables in two scopes:
+	// the inner class scope, and the outer one (whereever the class is declared).
+	// This map holds the outer ones to be later processed when the inner one is encountered.
+	// For why this is not a eslint issue see https://github.com/eslint/eslint-scope/issues/48#issuecomment-464358754
+	const identifierToOuterClassVariable = new WeakMap();
+
+	const checkPossiblyWeirdClassVariable = variable => {
+		if (isClassVariable(variable)) {
+			if (variable.scope.type === 'class') { // The inner class variable
+				const [definition] = variable.defs;
+				const outerClassVariable = identifierToOuterClassVariable.get(definition.name);
+
+				if (!outerClassVariable) {
+					return checkVariable(variable);
+				}
+
+				// Create a normal-looking variable (like a `var` or a `function`)
+				// For which a single `variable` holds all references, unline with `class`
+				const combinedReferencesVariable = {
+					name: variable.name,
+					scope: variable.scope,
+					defs: variable.defs,
+					identifiers: variable.identifiers,
+					references: variable.references.concat(outerClassVariable.references)
+				};
+
+				// Call the common checker with the newly forged normalized class variable
+				return checkVariable(combinedReferencesVariable);
+			}
+
+			// The outer class variable, we save it for later, when it's inner counterpart is encountered
+			const [definition] = variable.defs;
+			identifierToOuterClassVariable.set(definition.name, variable);
+
+			return;
+		}
+
+		return checkVariable(variable);
+	};
+
+	const checkVariable = variable => {
+		const variableReplacements = getVariableReplacements(replacements, variable);
+
+		if (variableReplacements.length === 0) {
+			return;
+		}
+
+		if (variable.defs.length === 0) {
+			return;
+		}
+
+		const scopes = variable.references.map(reference => reference.from).concat(variable.scope);
+
+		if (collideWithArgumentsSpecial(variableReplacements, scopes)) {
+			return;
+		}
+
+		const [definition] = variable.defs;
+
+		const problem = {
+			node: definition.name,
+			message: formatMessage(definition.name, variableReplacements)
+		};
+
+		if (variableReplacements.length === 1 && shouldFix(variable)) {
+			const [replacement] = variableReplacements;
+			const captureAvoidingReplacement = avoidCapture(replacement, scopes);
+
+			problem.fix = fixer => {
+				return variableIdentifiers(variable)
+					.map(fixIdentifier(fixer, captureAvoidingReplacement));
+			};
+		}
+
+		context.report(problem);
+	};
+
+	const checkVariables = scope => {
+		scope.variables.forEach(checkPossiblyWeirdClassVariable);
+	};
+
+	const checkChildScopes = scope => {
+		scope.childScopes.forEach(checkScope);
+	};
+
+	const checkScope = scope => {
+		checkVariables(scope);
+
+		return checkChildScopes(scope);
+	};
+
+	return {
+		'Program:exit'() {
+			checkScope(context.getScope());
+		}
+	};
+};
+
+const schema = [{
```

Comment:
If you upgrade to latest ESLint, we can now set default values in the schema. Not sure whether that's useful here. Just wanted to mention it. https://github.com/sindresorhus/eslint-plugin-unicorn/issues/247

Reply:
We'll have to up the eslint version in `peerDependencies`, let's not do this in this PR.